pbm/storage/s3/s3.go

```diff
@@ -397,182 +397,287 @@ func (s *S3) FileStat(name string) (inf storage.FileInfo, err error) {
-	s3obj, err := pr.sess.GetObject(&s3.GetObjectInput{
-		Bucket: aws.String(pr.opts.Bucket),
-		Key:    aws.String(path.Join(pr.opts.Prefix, pr.fname)),
-		Range:  aws.String(fmt.Sprintf("bytes=%d-%d", pr.n, pr.n+downloadChuckSize-1)),
-	})
-	if err != nil {
-		// if object size is undefined, we would read
-		// until HTTP code 416 (Requested Range Not Satisfiable)
-		var er awserr.RequestFailure
-		if errors.As(err, &er) && er.StatusCode() == http.StatusRequestedRangeNotSatisfiable {
-			return 0, io.EOF
+	go func() {
+		pr := s.newPartReader(name, fstat.Size, downloadChuckSize)
+		pr.Run(runtime.NumCPU())
+
+		defer func() {
+			w.Close()
+			pr.Reset()
+		}()
+
+		cbuf := &chunksBuf{}
+		heap.Init(cbuf)
+
+		for {
+			select {
+			case rs := <-pr.resultq:
+				// Although chunks are requested concurrently they must be written squentially
+				// to the destination as it's not necessary a file (decompress, mongorestore etc.).
+				// So if it's not a turn (previous chunks wasn't written yet) the chunk will be
+				// added to the buffer to wait.
+				if rs.meta.start != pr.written {
+					if len(*cbuf) <= buffMaxSize {
+						heap.Push(cbuf, &rs)
+					} else {
+						s.log.Warning("buffer is full (%d), reschedule part %d-%d", len(*cbuf), rs.meta.start, rs.meta.end)
+						go func() { pr.resultq <- rs }()
+					}
+					continue
+				}
+
+				err := pr.writeChunk(&rs, w, downloadRetries)
+				if err != nil {
+					w.CloseWithError(errors.Wrapf(err, "copy bytes %d-%d from resoponse buffer", rs.meta.start, rs.meta.end))
+					return
+				}
+
+				// check if we can send something from the buffer
+				for len(*cbuf) > 0 && []*chunk(*cbuf)[0].meta.start == pr.written {
+					r := heap.Pop(cbuf).(*chunk)
+					err := pr.writeChunk(r, w, downloadRetries)
+					if err != nil {
+						w.CloseWithError(errors.Wrapf(err, "copy bytes %d-%d from resoponse buffer", r.meta.start, r.meta.end))
+						return
+					}
+				}
+
+				// we've read all bytes in the object
+				if pr.written >= pr.fsize {
+					return
+				}
+
+			case err := <-pr.errc:
+				w.CloseWithError(errors.Wrapf(err, "download '%s/%s'", s.opts.Bucket, name))
+				return
+			}
 		}
-		pr.l.Warning("errGetObj Err: %v", err)
-		return 0, errGetObj(err)
-	}
-	if pr.tsize == -2 {
-		pr.setSize(s3obj)
-	}
+	}()
 
-	if pr.opts.ServerSideEncryption != nil {
-		sse := pr.opts.ServerSideEncryption
+	return r, nil
+}
 
-		s3obj.ServerSideEncryption = aws.String(sse.SseAlgorithm)
-		if sse.SseAlgorithm == s3.ServerSideEncryptionAwsKms {
-			s3obj.SSEKMSKeyId = aws.String(sse.KmsKeyID)
+func (pr *partReader) Run(concurrency int) {
+	pr.taskq = make(chan chunkMeta, concurrency)
+	pr.resultq = make(chan chunk)
+	pr.errc = make(chan error)
+	pr.close = make(chan struct{})
+
+	go func() {
+		for sent := int64(0); sent <= pr.fsize; {
+			select {
+			case <-pr.close:
+				return
+			default:
+			}
+			pr.taskq <- chunkMeta{sent, sent + downloadChuckSize - 1, 0}
```

Comment:
IMO it is better to make this sending as a select case. In that way, pr.close can stop at any time.
Here it could be blocked till another goroutine recv from the pr.taskq. Even if pr.close already closed